lib/shopify_api/graphql.ex

```diff
@@ -0,0 +1,119 @@
+defmodule ShopifyAPI.GraphQL do
+  @moduledoc """
+  Internal interface to Shopify's GraphQL Admin API, built on HTTPoison. 
+
+  Makes requests against Shopify GraphQL and returns a tuple containig
+  a %Response struct with %{response, metadata, status_code}
+
+  * `query/2`
+
+  iex> query = "mutation {
+    metafieldDelete (
+            input: {
+              id: \"gid://shopify/Metafield/1234567891234\"
+            }
+    )
+    {
+      deletedId
+      userErrors {
+        field
+        message
+      }
+      }
+  }"
+  iex> ShopifyAPI.GraphQL.query(auth, query)
+  {:ok, %Response{
+          response: %{
+            "metafield1" => %{
+              "deletedId" => "gid://shopify/Metafield/5256098316335"
+            }
+          },
+          metadata: %{
+            "cost" => %{
+              "actualQueryCost" => 10,
+              "requestedQueryCost" => 10,
+              "throttleStatus" => %{
+                "currentlyAvailable" => 990
+              }
+            }
+          }, 
+          status_code: 200
+        }
+  }
+
+  """
+
+  require Logger
+
+  alias ShopifyAPI.AuthToken
+  alias ShopifyAPI.GraphQL.{Response, Telemetry}
+
+  # Use HTTP in test for Bypass, HTTPS in all other environments
+  @transport if Mix.env() == :test, do: "http://", else: "https://"
+
+  @default_graphql_version "2019-07"
+
+  @log_module __MODULE__ |> to_string() |> String.trim_leading("Elixir.")
+
+  def query(%AuthToken{} = auth, query, opts \\ []) do
+    url = build_url(auth, opts)
+    headers = build_headers(auth, opts)
+
+    logged_request(auth, url, query, headers, opts)
+  end
+
+  def configured_version do
+    config = Application.get_env(:shopify_api, ShopifyAPI.GraphQL, [])
+    Keyword.get(config, :graphql_version, @default_graphql_version)
+  end
+
+  def remaining_points({:ok, %Response{metadata: metadata}}) do
+    metadata["cost"]["throttleStatus"]["currentlyAvailable"]
+  end
+
+  def remaining_points(_), do: nil
+
+  defp logged_request(auth, url, body, headers, options) do
+    {time, response} = :timer.tc(HTTPoison, :post, [url, body, headers, options])
+
+    response = Response.handle(response)
+
+    log_request(auth, response, time)
+
+    Telemetry.send(@log_module, auth, url, time, response)
+
+    response
+  end
+
+  defp log_request(%{app_name: app, shop_name: shop} = _token, response, time) do
+    Logger.debug(fn ->
+      status =
+        case response do
+          {:ok, %{status_code: status}} -> status
+          {:error, reason} -> "error[#{inspect(reason)}]"
+        end
+
+      "#{@log_module} for #{shop}:#{app} received #{status} points available (#{
+        remaining_points(response)
```

Comment:
This might be a good place to have a more direct breakdown of the request cost, as well as the overall rate limiting bucket. Since we have access to information here that we don't in REST responses, let's be sure to make the most of it for debugging.

A potentially useful format could be something like

```
[...] ShopifyAPI.GraphQL for example-shop:example-app received 200 in 123ms [cost 24, bucket 976/1000]
```

This may make the case for either a simpler method just returning the top-level map from the API response (ex `GraphQL.cost_extensions(%Response{})`) or one that does more work to unpack the rate limiting details from the response (ex `GraphQL.rate_limit_details(%Response{})`). With Shopify's API versioning giving us some degree of stability, I think the latter may be preferable.